test/Microsoft.Azure.Devices.Edge.Test.Common/config/BaseModuleConfigBuilder.cs

```diff
@@ -0,0 +1,141 @@
+            this.Env = new Dictionary<string, string>();
+            this.Name = name;
+            this.Settings = new Dictionary<string, object>()
+            {
+                ["image"] = image
+            };
+        }
+
+        public IModuleConfigBuilder WithEnvironment(IEnumerable<(string, string)> env)
+        {
+            foreach ((string key, string value) in env)
+            {
+                this.Env[key] = value; // for duplicate keys, last save wins!
+            }
+
+            return this;
+        }
+
+        public IModuleConfigBuilder WithProxy(Option<Uri> proxy, Protocol protocol)
+        {
+            proxy.ForEach(
+                p =>
+                {
+                    string proxyProtocol;
+                    switch (protocol)
+                    {
+                        case Protocol.Amqp:
+                            proxyProtocol = Protocol.AmqpWs.ToString();
+                            break;
+                        case Protocol.Mqtt:
+                            proxyProtocol = Protocol.MqttWs.ToString();
+                            break;
+                        case Protocol.AmqpWs:
+                        case Protocol.MqttWs:
+                            proxyProtocol = protocol.ToString();
+                            break;
+                        default:
+                            throw new ArgumentException("Unknown protocol");
+                    }
+
+                    // If UpstreamProtocol was already set in this config to a non-compatible value,
+                    // throw an error. The caller will need to fix the conflict in their code.
+                    // Otherwise, replace it with the proxy-compatible value.
+                    if (this.Env.TryGetValue("UpstreamProtocol", out string u))
+                    {
+                        if (u != protocol.ToString() && u != proxyProtocol)
+                        {
+                            string message = $"Setting \"UpstreamProtocol\" to \"{proxyProtocol}\"" +
+                                             $"would overwrite incompatible value \"{u}\"";
+                            throw new ArgumentException(message);
+                        }
+                    }
+
+                    this.WithEnvironment(
+                        new[]
+                        {
+                            ("https_proxy", p.ToString()),
+                            ("UpstreamProtocol", proxyProtocol)
+                        });
+                });
+
+            return this;
+        }
+
+        public IModuleConfigBuilder WithDesiredProperties(IDictionary<string, object> properties)
+        {
+            foreach ((string key, object value) in properties)
+            {
+                this.DesiredProperties[key] = value; // for duplicate keys, last save wins!
+            }
+
+            return this;
+        }
+
+        // How is module information from this class composed into a
```

Comment:
So looks like BaseModuleConfigBuilder also special cases EdgeAgent? Why so?
What about EdgeHub?

Reply:
Okay, I think I misunderstood the intent here.. I get it now.

Reply:
I'm not sure I understand what you're asking, but...

I saw two approaches for the builder interface. I could make it very dumb--you have to tell it _everything_ you want, which wouldn't be much different from just using object literal syntax to build the config JSON directly in the test. Or I could make it smarter--you tell me the modules you want in your deployment, and I'll do a lot of the boilerplate for you. Of course, since the config spec has some nesting--parts of the module config go down under $edgeAgent desired properties and part go into their own top-level object, the "smarter" interface needs to special-case edge agent. But with respect to the config spec, edge hub is a lot closer to a custom module so there's no special casing for it, other than adding it to `"systemModules"` instead of `"modules"`.